zookeeper/src/main/java/com/linecorp/armeria/client/zookeeper/ZooKeeperEndpointGroup.java

```diff
@@ -67,58 +61,40 @@ public ZooKeeperEndpointGroup(String zkConnectionStr, String zNodePath, int sess
+        ExponentialBackoffRetry retryPolicy = new ExponentialBackoffRetry(1000, 3);
+        CuratorFramework client = CuratorFrameworkFactory.builder()
+                                                         .connectString(zkConnectionStr)
+                                                         .retryPolicy(retryPolicy)
+                                                         .sessionTimeoutMs(sessionTimeout)
+                                                         .build();
+        client.start();
+        pathChildrenCache = new PathChildrenCache(client, zNodePath, true);
+        pathChildrenCache.getListenable().addListener((c, event) -> {
+            switch (event.getType()) {
+                case CHILD_ADDED:
+                    addEndpoint(nodeValueCodec.decode(event.getData().getData()));
+                    break;
+                case CHILD_REMOVED:
+                    removeEndpoint(nodeValueCodec.decode(event.getData().getData()));
+                    break;
+                default:
+                    break;
+            }
+        });
+        try {
+            pathChildrenCache.start();
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
+        }
     }
 
     @Override
     public void close() {
-        zooKeeperConnector.close(true);
-    }
-
-    /**
-     * Create a {@link ZooKeeperListener} listens specific ZooKeeper events.
-     * @return  {@link ZooKeeperListener}
-     */
-    private ZooKeeperListener createListener() {
-        return new ZooKeeperListener() {
-            @Override
-            public void nodeChildChange(Map<String, String> newChildrenValue) {
-                final List<Endpoint> newData = newChildrenValue.values().stream()
-                                                               .map(nodeValueCodec::decode)
-                                                               .filter(Objects::nonNull)
-                                                               .collect(toImmutableList());
-                final List<Endpoint> prevData = endpoints();
-                if (!prevData.equals(newData)) {
-                    setEndpoints(newData);
-                }
-            }
-
-            @Override
-            public void nodeValueChange(String newValue) {
-                //ignore value change event
-            }
-
-            @Override
-            public void connected() {
-            }
-        };
-    }
-
-    @VisibleForTesting
-    void enableStateRecording() {
-        zooKeeperConnector.enableStateRecording();
-    }
-
-    @VisibleForTesting
-    ZooKeeper underlyingClient() {
-        return zooKeeperConnector.underlyingClient();
-    }
-
-    @VisibleForTesting
-    BlockingQueue<KeeperState> stateQueue() {
-        return zooKeeperConnector.stateQueue();
+        try {
+            pathChildrenCache.close();
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
```

Comment:
If there's nothing much we can do after failing to close, how about just logging the failure?